src/ore/src/collections/hash.rs

```diff
@@ -0,0 +1,524 @@
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Stable wrappers around the stdlib `Hash` collections.
+//!
+//! The stdlib `Hash` collections (i.e., [`std::collections::HashMap`] and
+//! [`std::collections::HashSet`]) don't guarantee a stable iteration order. That is, the order in
+//! which elements are returned by methods like `iter` or `keys`, or in which elements are visited
+//! by methods like `retain`, is undefined and can differ between program executions. This property
+//! leads to non-determinism, which in turn can be the cause of bugs in logic that relies on
+//! deterministic execution.
+//!
+//! Since bugs caused by unstable iteration order are hard to spot, and it is not always easy to
+//! determine which parts of our code require determinism, we resort to banning the use of the
+//! stdlib `Hash` collections in general. Alternatives are available:
+//!
+//!  1. In most cases, you can simply switch to using the equivalent `BTree` collection type (i.e.,
+//!     [`std::collections::BTreeMap`] or [`std::collections::BTreeSet`]) as a drop-in replacement.
+//!     The `BTree` collections guarantee a stable iteration order, based on the `Ord` relation.
+//!  2. If the types you want to store don't (and shouldn't) implement `Ord`, or if profiling
+//!     reveals that the `BTree` collections are not suitable, use the wrappers provided in this
+//!     module instead.
+//!
+//! The `Hash` collection wrappers provided in this module only re-export methods that don't expose
+//! iteration order, and are therefore safe to use in code that relies on determinism.
+//!
+//! There are cases where the above mentioned alternatives are not sufficient, for example, if a
+//! third-party API requires you to pass a `HashMap` value. In this case, you can disable the
+//! clippy lint for the affected piece of code. Note that the onus is on you then to verify that
+//! doing so is sound and does not introduce bugs.
+
+use std::borrow::Borrow;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap as StdMap;
+use std::collections::HashSet as StdSet;
+use std::collections::TryReserveError;
+use std::hash::Hash;
+use std::ops::BitAnd;
+use std::ops::BitOr;
+use std::ops::BitXor;
+use std::ops::Index;
+use std::ops::Sub;
+
+/// A wrapper around [`std::collections::HashMap`] that hides methods that expose unstable
+/// iteration order.
+///
+/// See the module documentation for a rationale.
+#[derive(Clone, Debug, Default)]
+pub struct HashMap<K, V>(StdMap<K, V>);
```

Comment:
`#[repr(transparent)]` will guarantee this has the exact same layout!

Reply:
That seems like not a bad guarantee to have. Added!